qutip/tests/test_random.py

```diff
@@ -35,34 +35,145 @@
 from qutip import (
     rand_ket, rand_dm, rand_herm, rand_unitary, rand_ket_haar, rand_dm_hs,
     rand_super, rand_unitary_haar, rand_dm_ginibre, rand_super_bcsz, qeye,
+    rand_stochastic,
 )
 import pytest
 
 
-def test_rand_unitary_haar_unitarity():
+@pytest.mark.repeat(5)
+@pytest.mark.parametrize('func', [rand_unitary, rand_unitary_haar])
+def test_rand_unitary(func):
     """
     Random Qobjs: Tests that unitaries are actually unitary.
     """
-    U = rand_unitary_haar(5)
+    random_qobj = func(5)
     I = qeye(5)
-    assert U * U.dag() == I
+    assert random_qobj * random_qobj.dag() == I
 
 
+@pytest.mark.repeat(5)
+@pytest.mark.parametrize('density', [0.2, 0.8], ids=["sparse", "dense"])
+@pytest.mark.parametrize('pos_def', [True, False])
+def test_rand_herm(density, pos_def):
+    """
+    Random Qobjs: Hermitian matrix
+    """
+    random_qobj = rand_herm(5, density=density, pos_def=pos_def)
+    if pos_def:
+        assert all(random_qobj.eigenenergies() > 0)
+    assert random_qobj.isherm
+
+
+@pytest.mark.repeat(5)
+def test_rand_dm_Eigs():
+    """
+    Random Qobjs: Hermitian matrix - Eigs given
+    """
+    eigs = np.random.random(5)
+    eigs /= np.sum(eigs)
+    eigs.sort()
+    random_qobj = rand_herm(eigs)
+    assert np.allclose(random_qobj.eigenenergies(), eigs)
+    # verify hermitian
+    assert random_qobj.isherm
+
+
+@pytest.mark.repeat(5)
+@pytest.mark.parametrize('func', [rand_dm, rand_dm_hs])
+def test_rand_dm(func):
+    """
+    Random Qobjs: Density matrix
+    """
+    random_qobj = func(5)
+    assert abs(random_qobj.tr() - 1.0) < 1e-15
```

Comment:
can also be spelled `assert random_qobj.tr() == pytest.approx(1, abs=1e-15)` if you like, but I don't really mind.  My C upbringing makes me equally happy with the `abs(x - y) < tol` idiom for fp comparisons - don't know what other people think.